### PR TITLE
Improve error message when update cannot be installed due to version mismatch

### LINKED ISSUE #15259: Display more details on the package_error::app_version UI popup
The reasoning behind this change is that users do not understand the current error message and that they need to install previous updates before the one they're currently trying to install

### Current behaviour

image

---

### Wanted behaviour

The following package cannot be installed of top of the current game version: 
PACKAGE_NAME.pkg

**Installed version:** XX.XX
**Required version:** YY.YY
**Package version:** ZZ.ZZ

The package with the required version has to be installed before this one

### PR DESCRIPTION
**Pull request by MSuih**
Should hopefully make error messages easier to decipher for end users, and maybe also helps people trying to give help for those issues.

Before:
image

After:
image

Closes #15259

### CODE REVIEW

**Comment by AniLeo** [comment]
#15259

File: rpcs3/Crypto/unpkg.h [hunk 1]
```diff
@@ -297,11 +297,19 @@ struct PKGMetaData
...
 {
-	no_error,
-	app_version,
-	other
+	enum class error_type
+	{
+		no_error,
+		app_version,
+		other
+	} error;
```
Comment (Megamouse) [inline rpcs3/Crypto/unpkg.h:307]:
```suggestion
	} error = error_type::no_error;
```

File: rpcs3/Emu/system_utils.cpp [hunk 2]
```diff
@@ -81,8 +81,8 @@ namespace rpcs3::utils
 		named_thread worker("PKG Installer", [&]
 		{
 			std::deque<std::string> bootables;
-			const package_error error = package_reader::extract_data(reader, bootables);
-			return error == package_error::no_error;
+			const package_install_result error = package_reader::extract_data(reader, bootables);
+			return error.error == package_install_result::error_type::no_error;
```
Comment (Megamouse) [inline rpcs3/Emu/system_utils.cpp:85]:
```suggestion
			const package_install_result result = package_reader::extract_data(reader, bootables);
			return result.error == package_install_result::error_type::no_error;
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 3]
```diff
@@ -1191,10 +1191,28 @@ bool main_window::HandlePackageInstallation(QStringList file_paths, bool from_bo
 
 			ensure(package);
 
-			if (error == package_error::app_version)
+			if (result.error == package_install_result::error_type::app_version)
 			{
 				gui_log.error("Cannot install %s.", package->path);
-				QMessageBox::warning(this, tr("Warning!"), tr("The following package cannot be installed on top of the current data:\n%1!").arg(package->path));
+				bool hasExpected = !result.version.expected.empty();
+				bool hasFound = !result.version.expected.empty();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:1198]:
```suggestion
				const bool has_expected = !result.version.expected.empty();
				const bool has_found = !result.version.found.empty();
```

File: rpcs3/Crypto/unpkg.h [hunk 4]
```diff
@@ -297,11 +297,19 @@ struct PKGMetaData
...
+	struct version
+	{
+		std::string_view expected;
+		std::string_view found;
```
Comment (Megamouse) [inline rpcs3/Crypto/unpkg.h:311]: Never store temporary data in string_view
```suggestion
		std::string expected;
		std::string found;
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 5]
```diff
@@ -1191,10 +1191,28 @@ bool main_window::HandlePackageInstallation(QStringList file_paths, bool from_bo
...
+				if (hasExpected && hasFound)
+				{
+					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for version %1, but you have version %2.\n\nTried to install: %3")
+							.arg(QString::fromStdString(result.version.expected.data())).arg(QString::fromStdString(result.version.found.data())).arg(package->path));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:1202]:
```suggestion
					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for version %1, but you have version %2.\n\nTried to install: %3")
							.arg(QString::fromStdString(result.version.expected)).arg(QString::fromStdString(result.version.found)).arg(package->path));
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 6]
```diff
@@ -1191,10 +1191,28 @@ bool main_window::HandlePackageInstallation(QStringList file_paths, bool from_bo
...
+				}
+				else if (hasExpected)
+				{
+					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for version %1, but you don't have any data installed.\n\nTried to install: %3")
+							.arg(QString::fromStdString(result.version.expected.data())).arg(package->path));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:1207]:
```suggestion
					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for version %1, but you don't have any data installed.\n\nTried to install: %2")
							.arg(QString::fromStdString(result.version.expected)).arg(package->path));
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 7]
```diff
@@ -1191,10 +1191,28 @@ bool main_window::HandlePackageInstallation(QStringList file_paths, bool from_bo
...
+				}
+				else
+				{
+					// probably unreachable
+					QString found = hasFound ? tr("version %1").arg(QString::fromStdString(result.version.found.data())) : tr("no data installed");
+					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for unknown version, but you have version %2.\n\nTried to install: %3")
+							.arg(QString::fromStdString(result.version.expected.data())).arg(found).arg(package->path));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:1214]:
```suggestion
					const QString found = hasFound ? tr("version %1").arg(QString::fromStdString(result.version.found)) : tr("no data installed");
					QMessageBox::warning(this, tr("Warning!"), tr("Package cannot be installed on top of the current data.\nUpdate is for unknown version, but you have version %1.\n\nTried to install: %2")
							.arg(QString::fromStdString(result.version.expected)).arg(found).arg(package->path));
```

File: rpcs3/Crypto/unpkg.cpp [hunk 8]
```diff
@@ -1197,31 +1197,31 @@ package_error package_reader::extract_data(std::deque<package_reader>& readers,
...
+		if (error != package_install_result::error_type::no_error || num_failures > 0)
 		{
 			ensure(reader.m_result == result::error || reader.m_result == result::error_dirty);
-			return error;
+			return {error};
 		}
 
 		// Check if this package is allowed to be installed on top of the existing data
-		error = reader.check_target_app_version();
+		auto version_check = reader.check_target_app_version();
```
Comment (Megamouse) [inline rpcs3/Crypto/unpkg.cpp:1207]:
```suggestion
		const package_install_result version_check = reader.check_target_app_version();
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 9]
```diff
@@ -1191,10 +1191,28 @@ bool main_window::HandlePackageInstallation(QStringList file_paths, bool from_bo
 
 			ensure(package);
 
-			if (error == package_error::app_version)
+			if (result.error == package_install_result::error_type::app_version)
 			{
 				gui_log.error("Cannot install %s.", package->path);
-				QMessageBox::warning(this, tr("Warning!"), tr("The following package cannot be installed on top of the current data:\n%1!").arg(package->path));
+				bool has_expected = !result.version.expected.empty();
+				bool has_found = !result.version.found.empty();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:1198]:
```suggestion
				const bool has_expected = !result.version.expected.empty();
				const bool has_found = !result.version.found.empty();
```

**Pull request merged**